xlators/cluster/dht/src/dht-layout.c

```diff
@@ -55,41 +55,58 @@ dht_layout_t *
 dht_layout_get(xlator_t *this, inode_t *inode)
 {
     dht_layout_t *layout = NULL;
-    int ret = 0;
 
-    ret = dht_inode_ctx_layout_get(inode, this, &layout);
-    if ((!ret) && layout) {
-        GF_ATOMIC_INC(layout->ref);
-    }
+    dht_inode_ctx_layout_get(inode, this, &layout, 1);
+
     return layout;
 }
 
 int
 dht_layout_set(xlator_t *this, inode_t *inode, dht_layout_t *layout)
 {
     dht_conf_t *conf = NULL;
-    int oldret = -1;
     int ret = -1;
-    dht_layout_t *old_layout;
+    uint64_t ctx_int = 0;
+    dht_inode_ctx_t *ctx = NULL;
+    dht_layout_t *old_layout = NULL;
 
     conf = this->private;
     if (!conf || !layout)
         goto out;
 
-    LOCK(&conf->layout_lock);
+    LOCK(&inode->lock);
     {
-        oldret = dht_inode_ctx_layout_get(inode, this, &old_layout);
-        if (layout)
+        ret = __inode_ctx_get(inode, this, &ctx_int);
+        if (!ret) {
+            ctx = (dht_inode_ctx_t *)(uintptr_t)ctx_int;
+            if (ctx && ctx->layout) {
+                old_layout = ctx->layout;
+            }
+        }
+
+        if (ctx) {
+            ctx->layout = layout;
+        } else {
+            ctx = GF_CALLOC(1, sizeof(*ctx), gf_dht_mt_inode_ctx_t);
+            if (!ctx) {
+                ret = -1;
+            } else {
+                ctx->layout = layout;
+                ctx_int = (long)ctx;
+                ret = __inode_ctx_set0(inode, this, &ctx_int);
+            }
+        }
```

Comment:
Can't this be merged with `dht_inode_ctx_layout_set()` ? maybe if that function returns the old layout, if any, most of the common code can be avoided and this function can handle it correctly.

Reply:
Yes we can

Reply:
done